### PR TITLE
Add namespace name to OnMessage exception

### LINKED ISSUE #94: Namespace name in the OnMessage exception handler
Creating a new issue from part of #34.

Verbatim from @seanfeldman:
>When OnMessageOptions.ExceptionReceived event handler is invoked, it doesn't provide information on what namespace it's using. Having that information in the new client implementation would help for custom fail-over implementation.

### PR DESCRIPTION
**Pull request by SeanFeldman**
Fixes #94 

@vinaysurya please review

### CODE REVIEW

**Comment by SeanFeldman** [comment]
@jtaubensee you might be interested in this one. This fixes the left out issue from #34.

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs [hunk 1]
```diff
@@ -176,7 +179,7 @@ void CancelAutoRenewlock(object state)
             }
             catch (Exception exception)
             {
-                this.registerHandlerOptions.RaiseExceptionReceived(new ExceptionReceivedEventArgs(exception, ExceptionReceivedEventArgsAction.Complete));
+                this.registerHandlerOptions.RaiseExceptionReceived(new ExceptionReceivedEventArgs(exception, ExceptionReceivedEventArgsAction.Complete, namespaceName));
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: this. ?

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 2]
```diff
@@ -1101,7 +1101,7 @@ void ThrowIfNotPeekLockMode()
                 }
 
                 this.receivePumpCancellationTokenSource = new CancellationTokenSource();
-                this.receivePump = new MessageReceivePump(this, registerHandlerOptions, callback, this.receivePumpCancellationTokenSource.Token);
+                this.receivePump = new MessageReceivePump(this, registerHandlerOptions, callback, ServiceBusConnection.NamespaceName, this.receivePumpCancellationTokenSource.Token);
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: >ServiceBusConnection [](start = 98, length = 20)

Again, this. here

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs [hunk 3]
```diff
@@ -76,7 +79,7 @@ bool ShouldRenewLock()
                 catch (Exception exception)
                 {
                     MessagingEventSource.Log.MessageReceivePumpTaskException(this.messageReceiver.ClientId, string.Empty, exception);
-                    this.registerHandlerOptions.RaiseExceptionReceived(new ExceptionReceivedEventArgs(exception, ExceptionReceivedEventArgsAction.Receive));
+                    this.registerHandlerOptions.RaiseExceptionReceived(new ExceptionReceivedEventArgs(exception, ExceptionReceivedEventArgsAction.Receive, namespaceName));
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: if we are passing namespace, I feel we should definitely pass the entity name as well. Could you include `messageReceiver.Path` as well

**Comment by vinaysurya** [comment]
@SeanFeldman, thanks for taking this up! the changes look good to me. One thing  I was thinking though is would it be weird to expose only a namespace name on a entitylevel exception. What if someone comesback and says, I also need to know what entity name ran into the exception. So I was thinking how about exposing both as part of the args. that way users can have full control

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs [hunk 4]
```diff
@@ -15,28 +15,19 @@ sealed class SessionReceivePump
...
         readonly SemaphoreSlim maxPendingAcceptSessionsSemaphoreSlim;
 
-        public SessionReceivePump(
-            string clientId,
-            IMessageSessionEntity client,
-            ReceiveMode receiveMode,
-            SessionHandlerOptions sessionHandlerOptions,
-            Func<IMessageSession, Message, CancellationToken, Task> callback,
-            CancellationToken token)
+        public SessionReceivePump(string clientId, IMessageSessionEntity client, ReceiveMode receiveMode, SessionHandlerOptions sessionHandlerOptions, Func<IMessageSession, Message, CancellationToken, Task> callback, string namespaceName, CancellationToken token)
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: Could you revert this change? This is a long line and it is more readable with line breaks in between.

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs [hunk 5]
```diff
@@ -323,7 +314,7 @@ void RaiseExceptionRecieved(Exception e, string action)
                     // Lets not bother user with this exception.
                     if (!(exception is TaskCanceledException))
                     {
-                        this.sessionHandlerOptions.RaiseExceptionReceived(new ExceptionReceivedEventArgs(exception, ExceptionReceivedEventArgsAction.RenewLock));
+                        this.sessionHandlerOptions.RaiseExceptionReceived(new ExceptionReceivedEventArgs(exception, ExceptionReceivedEventArgsAction.RenewLock, namespaceName));
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: I think there are other places which you need to cover. For instance `CloseSessionIfNeededAsync()` method. Do take a second look at the usage of `RaiseExceptionReceived`

File: test/Microsoft.Azure.ServiceBus.UnitTests/Primitives/ServiceBusNamespaceConnectionTests.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Microsoft.Azure.ServiceBus.UnitTests.Primitives
+{
+    using Xunit;
+    public class ServiceBusNamespaceConnectionTests
+    {
+        [Fact]
+        public void Can_return_namespace_name_for_connection_string()
+        {
+            var namespaceConnection = new ServiceBusNamespaceConnection("Endpoint=sb://seanfeldman-test.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=7ry17m@yb31tw1llw0rk=");
```
Comment (nemakam) [inline test/Microsoft.Azure.ServiceBus.UnitTests/Primitives/ServiceBusNamespaceConnectionTests.cs]: 😈

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs (hunk 1 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: I really think `this.` is a noise and wish we'd move away from it 😞
But for this PR will do.

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs (hunk 3 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: In that case I'll introduce some Context object rather than multiple parameters. It will also be better from a public API perspective as signature on the method won't be considered broken if additions are made.

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs (hunk 5 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: `CloseSessionIfNeededAsync()` invokes `RaiseExceptionReceived()` which raises `ExceptionReceivedEventArgs` in a centralized manner, appending to the passed in information the namespace name @nemakam . Or am I missing something else?

File: test/Microsoft.Azure.ServiceBus.UnitTests/Primitives/ServiceBusNamespaceConnectionTests.cs (hunk 6 above)
Comment (SeanFeldman) [inline test/Microsoft.Azure.ServiceBus.UnitTests/Primitives/ServiceBusNamespaceConnectionTests.cs]: image

**Comment by SeanFeldman** [comment]
@vinaysurya @nemakam I've incorporated all the feedback and implemented the `ExceptionReceivedContext`. Please review.

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs [hunk 7]
```diff
@@ -12,20 +12,25 @@ namespace Microsoft.Azure.ServiceBus
...
+            Func<Message, CancellationToken, Task> callback, 
+            string namespaceName, 
+            string entityPath, 
             CancellationToken pumpCancellationToken)
         {
             this.messageReceiver = messageReceiver ?? throw new ArgumentNullException(nameof(messageReceiver));
             this.registerHandlerOptions = registerHandlerOptions;
             this.onMessageCallback = callback;
+            this.namespaceName = namespaceName;
+            this.entityPath = entityPath;
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: >entityPath [](start = 17, length = 10)

this.messageReceiver exposes a 'Path' already. we could simply use that?

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs [hunk 8]
```diff
@@ -115,7 +120,7 @@ bool ShouldRenewLock()
             catch (Exception exception)
             {
                 MessagingEventSource.Log.MessageReceiverPumpUserCallbackException(this.messageReceiver.ClientId, message, exception);
-                this.registerHandlerOptions.RaiseExceptionReceived(new ExceptionReceivedEventArgs(exception, ExceptionReceivedEventArgsAction.UserCallback));
+                this.registerHandlerOptions.RaiseExceptionReceived(new ExceptionReceivedEventArgs(exception, ExceptionReceivedEventArgsAction.UserCallback, this.namespaceName, this.entityPath));
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: >this.entityPath [](start = 176, length = 15)

this.messageReceiver.Path should work

File: src/Microsoft.Azure.ServiceBus/SessionPumpHost.cs [hunk 9]
```diff
@@ -13,13 +13,17 @@ internal sealed class SessionPumpHost
         readonly object syncLock;
         SessionReceivePump sessionReceivePump;
         CancellationTokenSource sessionPumpCancellationTokenSource;
+        readonly string namespaceName;
+        readonly string entityPath;
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/SessionPumpHost.cs]: >entityPath [](start = 24, length = 10)

potentially add this to the IMessageSessionEntity interface like we do in the IReceiverClient.

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs (hunk 5 above)
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: Nope... Not missing anything 🥇

**Comment by nemakam** [comment]
I'm just adding on-hold label for now so that we don't check this in. There are some small changes in the requirements of this task. 
We are planning to move out of 'namespace' model, to a complete 'endpoint' based connection. Watch out for a PR on this shortly that explains it.

**Comment by nemakam** [comment]
@SeanFeldman , refer #196.
Based on that, you might have to update your changes to give out endpoint details instead of namespace name

**Comment by SeanFeldman** [comment]
Thanks @nemakam for heads up. I'll watch that PR.

File: test/Microsoft.Azure.ServiceBus.UnitTests/Primitives/ServiceBusNamespaceConnectionTests.cs (hunk 6 above)
Comment (nemakam) [inline test/Microsoft.Azure.ServiceBus.UnitTests/Primitives/ServiceBusNamespaceConnectionTests.cs]: @SeanFeldman , keeping in view the public naming entities we use, could you rename seanfeldman to maybe 'contoso' or 'fabrikam'? Your pick!

Comment (SeanFeldman) [inline test/Microsoft.Azure.ServiceBus.UnitTests/Primitives/ServiceBusNamespaceConnectionTests.cs]: My pick is `land-of-oz` 😃

**Comment by SeanFeldman** [comment]
While #196 is brewing, have incorporated your feedback @vinaysurya @nemakam 
Blocked until #196 makes it into dev.

**Comment by SeanFeldman** [comment]
@vinaysurya are you sure to approve this? It's still marked as `on hold` and needs to incorporate changes from #196

**Comment by vinaysurya** [comment]
That is fine, well merge it only when Neerajs changes are checked in and
after your change on top of that.

**Comment by SeanFeldman** [comment]
Ok. I'm used to a flow when approved PR is a PR that can be merged 🙂

File: src/Microsoft.Azure.ServiceBus/Primitives/ServiceBusConnection.cs [hunk 10]
```diff
@@ -100,5 +100,9 @@ static void CloseConnection(AmqpConnection connection)
 
             return connection;
         }
+
+        /// <summary>Namespace name used for <see cref="ServiceBusConnection"/>.</summary>
+        /// <value>Namespace name.</value>
+        internal string NamespaceName => this.Endpoint.Host.Split(new[] {'.'}, StringSplitOptions.RemoveEmptyEntries)[0];
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Primitives/ServiceBusConnection.cs]: Based on #196, you'll have to change this to show the entire endpoint and not just the namespace... :)

Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Primitives/ServiceBusConnection.cs]: So for a
```
"Endpoint=sb://land-of-oz.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=7ry17m@yb31tw1llw0rk="
```

Exception context would surface `land-of-oz.servicebus.windows.net` @nemakam ?

Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Primitives/ServiceBusConnection.cs]: Yes. That's right.
We have public domains as well as separate national clouds. The domain name `servicebus.windows.net` should not be assumed.

Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Primitives/ServiceBusConnection.cs]: @nemakam are those (national clouds) documented anywhere?

Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Primitives/ServiceBusConnection.cs]: http://lmgtfy.com/?q=microsoft+azure+national+clouds 😄

Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Primitives/ServiceBusConnection.cs]: Haha, @nemakam lmgtfu was a good one, but it still doesn't answer if those domain names are documented as list somewhere.

Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Primitives/ServiceBusConnection.cs]: I don't know if different domain names are documented anywhere, but the domain names are definitely different. Also, having it as endpoint instead of namespace just makes it 1% more future-proof.. :)

Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Primitives/ServiceBusConnection.cs]: Not questioning that. I just find it... peculiar that those domains are not documented. @sethmanheim any ideas if that's documented anywhere for ASB? Thank you.

**Comment by SeanFeldman** [comment]
When creating a `ServiceBusNamespaceConnection` with 
```
Endpoint=sb://land-of-oz.servicebus.windows.net/;....
```
getting back for `Endpoint`
```
amqps://land-of-oz.servicebus.windows.net
```

`ServiceBusConnectionStringBuilder` hardcodes `amqps` as endpoint`s scheme. Is that correct?

**Comment by nemakam** [comment]
@SeanFeldman , Endpoint is a Uri, so having a `UriScheme` (in this case `amqps`) is a requirement. Based on use-case, we need to fetch the required properties. So you could do `endpoint.Authority` to get `land-of-oz.servicebus.windows.net`

**Comment by SeanFeldman** [comment]
@nemakam correct. My question is why `UriScheme` is replaced, not why we need it 🙂

**Comment by nemakam** [comment]
@SeanFeldman Ideally we should be parsing it from connection string, but I'd say YAGNI. As of now, there is no plan of extending the library to run with anything other than amqp protocol.

**Comment by SeanFeldman** [comment]
Reason I've asked @nemakam, aside from failing test, is because users that would register with a connection string that would contain `sb://` as their URI scheme would expect to see that value back. Instead, they'll be getting `amqps://` which will be confusing.

**Comment by nemakam** [comment]
Fair point. I need to test and verify the effects of keeping the scheme as-is, and changing the scheme only when user doesn't mention the scheme. Will take a look at it later. 
And just a reminder, there's action item pending on this PR. You still need to change the namespace to `endpoint` everywhere.

**Comment by SeanFeldman** [comment]
@nemakam correct. I'm in the mids, but my tests are failing due to the question I've raised with the URI schema 🙂

**Comment by nemakam** [comment]
Lets not block this PR for the above concern. Please go ahead with the PR assuming the `amqps` uriScheme., and when the `Urischeme` is picked up, will update your test accordingly.

**Comment by SeanFeldman** [comment]
Perhaps to raise an issue and reference this PR's discussion? That way
context can be tracked down when work on the new issue is started.

**Comment by SeanFeldman** [comment]
@nemakam I've pushed the change.
Also raised an issue #197

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs [hunk 11]
```diff
@@ -12,20 +12,22 @@ namespace Microsoft.Azure.ServiceBus
     sealed class MessageReceivePump
     {
         readonly Func<Message, CancellationToken, Task> onMessageCallback;
+        readonly string namespaceName;
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: rename this to endpoint

File: src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
+namespace Microsoft.Azure.ServiceBus.Primitives
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: Add  copyright

File: src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    using System;
+
+    /// <summary>Context provided for <see cref="ExceptionReceivedEventArgs"/> exception raised by the client.</summary>
+    public class ExceptionReceivedContext
+    {
+        /// <summary>Initializes a new instance of the <see cref="ExceptionReceivedContext" /> class.</summary>
+        /// <param name="action">The action associated with the exception.</param>
+        /// <param name="endpoint">The endpoint associated with the exception.</param>
+        /// <param name="entityPath">The entity path associated with the exception.</param>
+        public ExceptionReceivedContext(string action, string endpoint, string entityPath)
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: internal constructor?

File: src/Microsoft.Azure.ServiceBus/QueueClient.cs [hunk 14]
```diff
@@ -197,7 +197,8 @@ internal SessionPumpHost SessionPumpHost
                             this.sessionPumpHost = new SessionPumpHost(
                                 this.ClientId,
                                 this.ReceiveMode,
-                                this.SessionClient);
+                                this.SessionClient,
+                                this.ServiceBusConnection.Endpoint.ToString());
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/QueueClient.cs]: Should this be `Endpoint.ToString()` or `Endpoint.Authority`?
Former would give out `amqps://abcd.servicebus.windows.net:80`, and latter would be `abcd.servicebus.windows.net`.
I feel the latter is cleaner. 
This will also relieve the need of #197

Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/QueueClient.cs]: Same with `MessageReceiver` and other classes

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs [hunk 15]
```diff
@@ -15,28 +15,22 @@ sealed class SessionReceivePump
...
         readonly SemaphoreSlim maxPendingAcceptSessionsSemaphoreSlim;
 
-        public SessionReceivePump(
-            string clientId,
-            IMessageSessionEntity client,
-            ReceiveMode receiveMode,
-            SessionHandlerOptions sessionHandlerOptions,
-            Func<IMessageSession, Message, CancellationToken, Task> callback,
-            CancellationToken token)
+        public SessionReceivePump(string clientId, IMessageSessionEntity client, ReceiveMode receiveMode, SessionHandlerOptions sessionHandlerOptions,
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: Could you revert this formatting and make it multiline again?

File: test/Microsoft.Azure.ServiceBus.UnitTests/Primitives/ServiceBusNamespaceConnectionTests.cs [hunk 16]
```diff
@@ -0,0 +1,34 @@
+﻿using Microsoft.Azure.ServiceBus.Primitives;
```
Comment (nemakam) [inline test/Microsoft.Azure.ServiceBus.UnitTests/Primitives/ServiceBusNamespaceConnectionTests.cs]: copyrights

Comment (nemakam) [inline test/Microsoft.Azure.ServiceBus.UnitTests/Primitives/ServiceBusNamespaceConnectionTests.cs]: Move using inside namespace

File: src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs (hunk 12 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: Need a bot to do so 😃

File: src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs (hunk 13 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: If someone want to perform testing and needs to construct the object to mock behavior, `internal` will get in a way. Do you still think it should be internal?

Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: makes sense. We could leave it `public`

File: src/Microsoft.Azure.ServiceBus/QueueClient.cs (hunk 14 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/QueueClient.cs]: I could do that. Would it be still called "endpoint"?
